synapse/rest/media/v1/preview_url_resource.py

```diff
@@ -367,7 +368,128 @@ async def _do_preview(self, url: str, user: UserID, ts: int) -> bytes:
+                url,
+                output_stream=output_stream,
+                max_size=self.max_spider_size,
+                headers={"Accept-Language": self.url_preview_accept_language},
+            )
+        except SynapseError:
+            # Pass SynapseErrors through directly, so that the servlet
+            # handler will return a SynapseError to the client instead of
+            # blank data or a 500.
+            raise
+        except DNSLookupError:
+            # DNS lookup returned no results
+            # Note: This will also be the case if one of the resolved IP
+            # addresses is blacklisted
+            raise SynapseError(
+                502,
+                "DNS resolution failure during URL preview generation",
+                Codes.UNKNOWN,
+            )
+        except Exception as e:
+            # FIXME: pass through 404s and other error messages nicely
+            logger.warning("Error downloading %s: %r", url, e)
+
+            raise SynapseError(
+                500,
+                "Failed to download content: %s"
+                % (traceback.format_exception_only(sys.exc_info()[0], e),),
+                Codes.UNKNOWN,
+            )
+
+        if b"Content-Type" in headers:
+            media_type = headers[b"Content-Type"][0].decode("ascii")
+        else:
+            media_type = "application/octet-stream"
+
+        download_name = get_filename_from_headers(headers)
+
+        # FIXME: we should calculate a proper expiration based on the
+        # Cache-Control and Expire headers.  But for now, assume 1 hour.
+        expires = ONE_HOUR
+        etag = headers[b"ETag"][0].decode("ascii") if b"ETag" in headers else None
+
+        return length, uri, code, media_type, download_name, expires, etag
+
+    async def _parse_data_url(
+        self, url: str, output_stream: BinaryIO
+    ) -> Tuple[int, str, int, str, Optional[str], int, Optional[str]]:
+        """
+        Parses a data: URL.
+
+        Args:
+             url: The URL to parse.
+             output_stream: The stream to write the content to.
+
+        Returns:
+            A tuple of:
+                Media length, URL downloaded, the HTTP response code,
+                the media type, the downloaded file name, the number of
+                milliseconds the result is valid for, the etag header.
+        """
+
+        try:
+            logger.debug("Trying to parse data url '%s'", url)
+            with urlopen(url) as url_info:
+                # TODO Can this be more efficient.
+                output_stream.write(url_info.read())
+        except Exception as e:
+            logger.warning("Error parsing data: URL %s: %r", url, e)
+
+            raise SynapseError(
+                500,
+                "Failed to parse data URL: %s"
+                % (traceback.format_exception_only(sys.exc_info()[0], e),),
+                Codes.UNKNOWN,
+            )
+
+        length = output_stream.seek(1)
```

Comment:
Not sure I follow this. Isn't this seeking the stream to the first byte of the output, and returning the integer `1`? (Judging by [docs](https://docs.python.org/3/library/io.html#io.IOBase.seek)). If so, we might be making the same mistake [here](https://github.com/matrix-org/synapse/blob/d3cf0730f8bf3de58d37060eff0dca3f69b4a0e1/tests/server.py#L316).

Should this be [`output_stream.tell()`](https://docs.python.org/3/library/io.html#io.IOBase.tell)?

Reply:
Yes, this should definitely be using `tell()`! 🎉

Reply:
I made the change here since it seems to just be a minor wart in our testing framework.

Reply:
I fixed this in b9a1c66814c3de9334c45ba7cd1dd64850ffa58c and added a test for it in d38a07ccbd0c6a2ba70ba3a8f6fde3a15bf0a4b6.